### DISCUSSION TITLE
Handling Trailing Slashes in Routes - Pocketbase 0.23.8

### DISCUSSION THREAD

**Discussion by rudyrdx** [pocketbase/pocketbase#6128] [Q&A] [answered]
Hi,  

I have been using PocketBase for 5 months and have successfully updated my apps to the latest version. During this process, I encountered an issue related to trailing slashes (`/`) in routes.  

### Issue  
From an API's perspective, having a trailing slash at the end of a route (e.g., `/hello/`) should technically be treated as a different route than one without it (e.g., `/hello`). However, this behavior may not always be desirable, especially if we want routes to resolve irrespective of the trailing slash.  

**Example:**  
```go
e.Router.GET("/hello", func(re *core.RequestEvent) error {
    return re.String(200, "Hello world!")
})
```
In the above example:  
- A GET request to `/hello` works as expected and returns `"Hello world!"`.  
- A GET request to `/hello/` results in a `resource not found` error.  

### My Attempted Solutions  

1. **Redirect Middleware**  
   I implemented middleware to check if a URL path ends with a trailing slash and redirect it to the route without the slash.  
   - While this works, the server has to respond twice (once for the redirect), which is not ideal.  
   - It also caused conflicts with the admin page, so I had to filter the admin page route in the middleware.

2. **Intercept and Adjust Route Before Not Found Error**  
   I attempted to create middleware with a high priority to intercept the request before it reaches the "not found" error. My goal was to check for a route without the trailing slash using the `Router.HasRoute()` method and update the path accordingly.

   **Code Example:**  
   ```go
   e.Router.Bind(&hook.Handler[*core.RequestEvent]{
       Id: "resolveAddr",
       Func: func(f *core.RequestEvent) error {
           path := f.Request.URL.Path
           fmt.Println(path)
           if strings.HasSuffix(path, "/") {
               routeWithoutSuffix := strings.TrimSuffix(path, "/")
               if e.Router.HasRoute(http.MethodGet, routeWithoutSuffix) {
                   f.Request.URL.Path = routeWithoutSuffix
               }
           }
           return e.Next()
           // doesn't point to the correct route
       },
       Priority: 999,
   })
   ```

   However, this did not resolve the issue as expected.  

Any suggestions on how I can solve this?

**Comment by ganigeorgiev** [comment]
PocketBase v0.23 router is a thin wrapper around the standard Go `net/http.ServeMux` and it follows the same rules for resolving the registered routes, aka. the ones described in https://pocketbase.io/docs/go-routing/#path-parameters-and-matching-rules.

Before creating a rewrite rule you should ask yourself whether this is something really worth the dev effort and needed in the first place. Personally for me, the answer would be No.

It seems that you want "strip/rewrite" type of middleware but this is not easily done without hijacking the connection (_~or at least I can't think of other option at the moment~ see comment https://github.com/pocketbase/pocketbase/discussions/6128#discussioncomment-11594735_) and I don't recommend it as it is very error prone.

The easiest workaround would be to use a 301 redirect in your middleware (_I'm not sure what do you mean by _"caused conflicts with the admin page"_ since I don't think the dashboard sends any requests to trailing slash API resources_).

Another option could be to register your path as `/hello/{$}`. This way the std Go mux should redirect `/hello` to `/hello/` for you.

Of course you can also register the routes twice if you want the user to remain on the specific visited path without redirect:

```go
// see the note in https://pkg.go.dev/net/http#hdr-Trailing_slash_redirection-ServeMux

fn := func(re *core.RequestEvent) error {
  return re.String(200, "Hello world!")
}
se.Router.GET("/hello", fn)
se.Router.GET("/hello/{$}", fn)
```

_As a side-note: the `Priority` field actually works the opposite way - the higher the number the later it executes, aka. it is like an "order" field (_maybe I should have named it differently but I don't think it is worth a breaking change now_)._

**Reply by rudyrdx** [reply]
Hi,

Thank you for taking the time to respond. I tried to understand the priority concept by the example in the Git repository and the accompanying comment.
https://github.com/pocketbase/pocketbase/blob/9b4200a65cff7ac7fc41f257b2eb66dfabe0faba/examples/base/main.go#L116

Regarding the 301 redirect issue I encountered, I had identified the problem to be a logical issue and nothing to do with pocketbase's implementation. I fixed it by adding a condition with `HasRoute()` method which at the time of writing the question I hadn't tried.
Just for the reference here's the code:
```go
se.Router.Bind(&hook.Handler[*core.RequestEvent]{
    Id: "ResolveSlash",
    Func: func(e *core.RequestEvent) error {
        path := e.Request.URL.Path
        path_without_slash := strings.TrimSuffix(path, "/")
        if strings.HasSuffix(path, "/") {
            return e.Redirect(http.StatusMovedPermanently, path_without_slash)
        } else {
            fmt.Println("route not found: ")
        }
        return e.Next()
    },
    Priority: 1,
})
se.Router.GET("/hello", func(re *core.RequestEvent) error {
    return re.String(200, "Hello world!")
})
```
Navigating to the `/_` route, this code causes `too many requests to server error`, but with the addition of the HasRoute() condition
```go
if strings.HasSuffix(path, "/") && se.Router.HasRoute(http.MethodGet, path_without_slash)
```
Fixes the `too many requests to server error`.

So for now I think I should go with the redirect method. But would also like to try and come up with a solution for myself for the trailing `/` problem. Could you point me towards some files in the source as a starting point?

Thankyou.

**Reply by ganigeorgiev** [reply, marked as answer]
**Without context why and what is actually the problem you are trying to solve, the above for me remains unnecessary.**

If you want both `/hello` and `/hello/` to resolve to a single route either simply register the route path as `/hello/{$}` or register it twice if you don't want redirect and want both path to resolve on their own (see the note in the official [go docs](https://pkg.go.dev/net/http#hdr-Trailing_slash_redirection-ServeMux)).

The standard Go ServeMux doens't allow changing the route matching rules on the fly because it kindof defeat the purpose of having the rules in the first place (_that is what the older pre Go v1.22 mux implementation basically was_).

Even the [current default PocketBase 404 handler](https://github.com/pocketbase/pocketbase/blob/9b4200a65cff7ac7fc41f257b2eb66dfabe0faba/tools/router/router.go#L62-L71) is a "hack" because at least as of now Go doesn't expose the builtin error handler (_see also https://github.com/golang/go/issues/65648_).

One option as explained above could be to take control entirely over the connection (_this is what I meant by "connection hijacking"; see the docs about the [`Hijacker` interface](https://pkg.go.dev/net/http#Hijacker) and its example_).

Another one if you really want to intercept and rewrite the path BEFORE the matching happens you'll have to somehow do it before the [`serveMux121.findHandler` function](https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/net/http/servemux121.go;l=104) call. In PocketBase one such option could be to wrap the `se.Server.Handler` field after `se.Next()` because the Go mux is assigned in the finalizer of the `OnServe` hook, for example:

```go
app.OnServe().BindFunc(func(se *core.ServeEvent) error {
  se.Router.GET("/hello", func(re *core.RequestEvent) error {
    return re.String(200, "Hello world!")
  })

  serveErr := se.Next()
  if serveErr != nil {
    return serveErr
  }

  // wrap the resulting main server handler
  // (no new routes can be registered via se.Router after se.Next())
  originalHandler := se.Server.Handler
  se.Server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    // do something with the path, like always stripping the trailing slash ...
    r.URL.Path = strings.TrimRight(r.URL.Path, "/")

    originalHandler.ServeHTTP(w, r)
  })

  return nil
})
```

**But again, don't use the above if you don't really have a good reason for.**
